shellcheck: add override for newer version

zowoq opened the pull request:
###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

cdepillabout commented on pkgs/development/haskell-modules/configuration-common.nix:
```diff
@@ -2364,5 +2364,8 @@ self: super: {
   } super.postgrest));
 
+  # build newer version for `pkgs.shellcheck`
+  ShellCheck_0_9_0 = doDistribute super.ShellCheck_0_9_0;
```
I don't think you actually need `doDistribute` here, since it will get built anyway from the top-level `shellcheck` depending on it.

zowoq replied:
It's needed otherwise hydraPlatforms is `[ ]` for `pkgs.shellcheck`.

cdepillabout replied:
Ah, I see.  So I guess `pkgs.shellcheck` just uses the same hydraPlatforms as `haskellPackages.ShellCheck`.  Makes sense.

cdepillabout commented on pkgs/top-level/all-packages.nix:
```diff
@@ -18071,7 +18071,8 @@ with pkgs;
   shellcheck = callPackage ../development/tools/shellcheck {
-    inherit (__splicedPackages.haskellPackages) ShellCheck;
+    # inherit (__splicedPackages.haskellPackages) ShellCheck;
+    ShellCheck = __splicedPackages.haskellPackages.ShellCheck_0_9_0;
```
I don't necessarily want to argue against this change, but in the past we have generally only provided Haskell packages at the versions locked in the LTS release we are tracking.  We generally don't provide top-level packages for Haskell package versions that are newer than in LTS.

We have an open issue about this: https://github.com/NixOS/nixpkgs/issues/190542

Is there some reason you think that ShellCheck should be "special" in this regard, and have the top-level package be the latest version (as opposed to whatever version is in the LTS)?  It generally creates more work for the Haskell team to have packages tracking a version not in the LTS (not that this is a super-important reason).

zowoq replied:
It used to be that `shellcheck` was always the newest version but that override was dropped for some reason. (IIRC it was somewhere in the haskell tooling outside this repo)

> Is there some reason you think that ShellCheck should be "special" in this regard, and have the top-level package be the latest version (as opposed to whatever version is in the LTS)?

TBH I think this is looking at it wrong. LTS tracking is only relevant because this happens to be written in haskell which is a detail users most likely don't care about. Is there a reason to not have the newest version of this  tool? 

I did the same PR a year ago: https://github.com/NixOS/nixpkgs/pull/147780

>  It generally creates more work for the Haskell team to have packages tracking a version not in the LTS ...

Is there any way to have `pkgs.shellcheck` be independent of `haskellPackages`, e.g switch to using a FOD for the deps?

cdepillabout replied:
> Is there any way to have pkgs.shellcheck be independent of haskellPackages, e.g switch to using a FOD for the deps?

We unfortunately don't have any way of doing this for the Haskell stuff in Nixpkgs right now.  (Although I agree this would potentially be a better solution for a package like shellcheck.  The top-level `purescript` package is somewhat similar, and if you take a look at the derivation you'll see that we're distributing patchelf'd binaries, instead of building from source.)

> Is there a reason to not have the newest version of this tool?

It causes slightly more work for the Haskell maintainers here in Nixpkgs.  When a new version of ShellCheck is released (like 0.9.1), `haskellPackages.ShellCheck_0_9_0` will no longer be available (instead we'll only have `haskellPackages.ShellCheck_0_9_1`), and one of us will have to manually update the reference here in `all-packages.nix`.  It creates extra work for us.

Not that I think that is necessarily a reason to prevent this PR from being merged in.

zowoq replied:
> ... distributing patchelf'd binaries ...

shellcheck upstream doesn't have binaries for many platforms so this probably isn't an option at the moment.

> > Is there a reason to not have the newest version of this tool?
> It causes slightly more work for the Haskell maintainers here ...

Yes, sorry I meant besides that. I am sympathetic to not wanting more work and that every little bit adds up but not providing the newest version of a dev tool seems a bit odd.

Also shellcheck doesn't release very often so likely the next time this needs to be touched will to remove it as 0.9.0 is the default.

cdepillabout replied:
> shellcheck upstream doesn't have binaries for many platforms so this probably isn't an option at the moment.

Ah, I was agreeing with you that not having some sort of FOD builder for Haskell is unfortunate, and the purescript situation is less-than-ideal!

> Also shellcheck doesn't release very often so likely the next time this needs to be touched will to remove it as 0.9.0 is the default.

That makes sense, and sounds like a good reason to go ahead with this change.

maralorn replied:
> It used to be that `shellcheck` was always the newest version but that override was dropped for some reason. (IIRC it was somewhere in the haskell tooling outside this repo)

Those overrides actually exist within nixpkgs and I cleaned them up a while ago. Not sure if shellcheck was in there at that point. Maybe we can introduce a policy that any maintainer of a package is allowed to enter their package there (with a comment).

zowoq replied:
> Those overrides actually exist within nixpkgs

Oh, thanks. I had looked through haskell-updates history but I forgot about the scripts in maintainers.

> Not sure if shellcheck was in there at that point.

Looks like it was: bf3a3b62e4a98d3b721ff222d7a457e3c013aaf5

> Maybe we can introduce a policy that any maintainer of a package is allowed to enter their package there ...

I'm not maintaining it currently but if this approach makes things easier I can add myself. My haskell knowledge is rather limited so I probably won't be of much use if the package breaks in a non-trivial way.

maralorn replied:
You have proven enough nix knowledge to be fully qualified for a maintainership. Deep Haskell knowledge is rarely required and if it is, you can always ask in #haskell:nixos.org.

SebTM commented:
I'm sorry to ask but I was about to make a PR myself (with copying the shellcheck in haskell-modules) and trying to figure out how this solution works/why as I can't find any reference on `ShellCheck_0_9_0` so what I found out so far it seems parsed like "<Package>_<Version>" but isn't there any hash to maintain to compare?

zowoq committed:
shellcheck, haskellPackages.ShellCheck: add maintainer

zowoq committed:
haskellPackages.ShellCheck: unpin

zowoq committed:
haskellPackages.haskell-ci: pin to ShellCheck_0_8_0

zowoq committed:
haskellPackages: regenerate package set based on current config

This commit has been generated by maintainers/scripts/haskell/regenerate-hackage-packages.sh

sternenseemann commented:
Hopefully everyone who needs a specific version of ShellCheck (i.e. need specific rules) pin as strictly as haskell-ci.

In favor as it mirrors how things have been historically.
